File: leetcode/Arrays/leetcode-46/46.go
package arrays

import "sort"

func permute(nums []int) [][]int {
	res := [][]int{}
	if nums == nil || len(nums) == 0 {
		return res
	}

	sort.Ints(nums)
	tmp := []int{}
	for _, v := range nums {
		tmp = append(tmp, v)
	}
	res = append(res, tmp)

	tmp = nextPermute(nums)
	for len(tmp) > 0 {
		res = append(res, tmp)
		tmp = nextPermute(nums)
	}
	return res
}

func nextPermute(nums []int) []int {
	res := []int{}
	if nums == nil || len(nums) == 0 {
		return res
	}

	firstSmall := -1
	for i := len(nums) - 2; i >= 0; i-- {
		if nums[i] < nums[i+1] {
			firstSmall = i
			break
		}
	}

	if firstSmall == -1 {
		return res
	}

	for i := len(nums) - 1; i > firstSmall; i-- {
		if nums[i] > nums[firstSmall] {
			nums[i], nums[firstSmall] = nums[firstSmall], nums[i]
			break
		}
	}

	i, j := firstSmall+1, len(nums)-1
	for i < j {
		nums[i], nums[j] = nums[j], nums[i]
		i++
		j--
	}

	for _, v := range nums {
		res = append(res, v)
	}
	return res
}

File: leetcode/Arrays/leetcode-46/46.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        if (nums == null || nums.length == 0) {
            return res;
        }

        Arrays.sort(nums);
        res.add(getList(nums));

        while (next_permute(nums)) {
            res.add(getList(nums));
        }
        return res;
    }

    private List<Integer> getList(int[] nums) {
        List<Integer> res = new ArrayList<>();
        if (nums == null || nums.length == 0) {
            return res;
        }

        for (int num : nums) {
            res.add(num);
        }
        return res;
    }

    private boolean next_permute(int[] nums) {
        if (nums == null || nums.length == 0) {
            return false;
        }

        int firstSmall = -1;
        for (int i = nums.length - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1]) {
                firstSmall = i;
                break;
            }
        }

        if (firstSmall == -1) {
            return false;
        }

        for (int i = nums.length - 1; i > firstSmall; i--) {
            if (nums[i] > nums[firstSmall]) {
                swap(nums, i, firstSmall);
                break;
            }
        }
        reverse(nums, firstSmall + 1, nums.length - 1);
        return true;
    }

    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

    private void reverse(int[] nums, int start, int end) {
        while (start < end) {
            swap(nums, start++, end--);
        }
    }
}